### PR TITLE
Bugfix to varentropy calculation 

### PR DESCRIPTION
**Pull request by nreHieW**
This PR does 2 things:
1) Adds dynamic torch device support (similar to #https://github.com/xjdr-alt/entropix/pull/3) but with the device logic only being handled in `torch_main.py`
2) More importantly, it fixes variance of attention entropy calculation (see below)

In `calculate_metrics()`, `attn_entropy` is of shape `(bs, heads,1)`. The variance computation is happening over `axix/dim = -1` which means its computing the variance over a single element which I'm guessing is unintended behaviour. 

`jnp.var` over a single item returns 0 so this basically means that `attn_varentropy` has no effect on `adaptive_sample`

The torch parts have a bigger problem. `torch.var` over a single item returns `nan` so `temperature` and `top p` are both `nan`. This means that the entire logit tensor is `nans`. Ordinarily, `torch.multinomial` cannot sample over a tensor with `nan` but `_sample` has a check for nans and fills them which means that in this case the tensor is a constant tensor so the sampling would be completely random 

The fix is pretty straightforward, compute the variance over the head dimension `dim = 1` instead (which I assume is the intended call).

### CODE REVIEW

**Comment by xjdr-alt** [comment]
Thank you for the PR!
